src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -293,3 +290,23 @@
       wrap-with-certificate-cn
       wrap-with-default-body
       wrap-with-debug-logging))
+
+(defn wrap-with-parent-check
+  "Middleware that checks the parent exists before serving the rest of the
+   application. This ensures we always return 404's on child paths when the
+   parent data is empty."
+  [app version parent id]
+  (fn [{:keys [globals] :as req}]
+    (let [{:keys [scf-read-db url-prefix]} globals
+          status-fn (case parent
```

Comment:
Could we make this function more generic, having it take an existence check function as an args, and simply wrap the individual handlers in the nodes/env/report endpoints? I'd like to be able to see what function is being used in the endpoints themselves